In [1]:
import pandas as pd
from collections import Counter
file_path = 'input_game.csv'
df = pd.read_csv(file_path)

def always_trust(actions):    
    counter_cheat=0  
    for action in actions:
        if(action!='TRUST'):
            counter_cheat+=1
    if(counter_cheat <0.1 * len(actions)):
        return True
    return False
def always_cheat(actions):    
    counter_trust=0  
    for action in actions:
        if(action!='CHEAT'):
            counter_trust+=1
    if(counter_trust <0.1 * len(actions)):
        return True
    return False
def copycat(actions,reaction):
    error=0
    for i in range(1,len(actions)):
        if actions[i]!=reaction[i-1]:
            error+=1
    if error<0.1*len(actions):
        return True
    else:
        return False
def rev_copycat(actions,reaction):
    error=0
    for i in range(1,len(actions)):
        if actions[i]==reaction[i-1]:
            error+=1
    if error<0.1*len(actions):
        return True
    else:
        return False
def simpleton(actions,reaction):
    error=0
    for i in range(1,len(actions)):
        if reaction[i-1]=='TRUST':
            if actions[i]!=actions[i-1]:
                error+=1
        else:
            if actions[i]==actions[i-1]:
                error+=1
                
    if error<0.1*len(actions):
        return True
    else:
        return False
def grudger(actions,reaction):
    error=0
    cheated=False
    if actions[0]=='CHEAT':
        error=1
    if reaction[0]=='CHEAT':
        cheated=True
    for i in range(1,len(actions)):
        if cheated==False:
            if reaction[i-1]=='TRUST' and actions[i]!='TRUST':
                error+=1
            if reaction[i]=='CHEAT':
                cheated=True
        elif cheated==True:
            if actions[i]!='CHEAT':
                error+=1
    if error<0.1*len(actions):
        return True
    else:
        return False
def Trust3Cheat1(actions):
    error=0
    i=0
    for action in actions:
        if i!=3 :
            if action!='TRUST':    
                error+=1
                i=-1
            i+=1
        else:
            if action!='CHEAT':
                error+=1
            i=0
    if error<0.1*len(actions):
        return True
    else:
        return False
def copykitten(actions,reaction):
    error=0
    if actions[0]=='CHEAT':
        error+=1
    if actions[1]=='CHEAT':
        error+=1    
    Moreprev = reaction[0]
    prev = reaction[1]
    for i in range(2,len(actions)):
        if prev == 'CHEAT' and Moreprev=='CHEAT':
            if actions[i]!='CHEAT':
                error+=1
        else:
            if actions[i]!='TRUST':
                error+=1
        Moreprev=prev
        prev=reaction[i]
    if error<0.1*len(actions):
        return True
    else:
        return False

def identify_strategy(actions,reaction):
    if always_trust(actions):
        return 'Always TRUST'
    elif always_cheat(actions):
        return 'Always CHEAT'
    elif rev_copycat(actions, reaction):
        return 'Reverse Copycat '
    elif copycat(actions, reaction):
        return 'CopyCat'
    elif simpleton(actions, reaction):
        return 'Simpleton'
    elif grudger(actions, reaction):
        return 'Grudger'
    elif copykitten(actions, reaction):
        return 'COPYkitten'
    elif Trust3Cheat1(actions):
        return 'Trust3 Cheat1'
    else:
        return 'Other'

# Apply the function to each player
player_strategies = {}
for player_id in df['p1_id'].unique():
    player_actions = df[df['p1_id'] == player_id]['p1_action'].tolist()
    opponent_actions = df[df['p1_id'] == player_id]['p2_action'].tolist()
    # print(len(player_actions))
    # print(len(opponent_actions))
    strategy = identify_strategy(player_actions,opponent_actions)
    player_strategies[player_id] = strategy

# Count the number of players following each strategy
strategy_counts = Counter(player_strategies.values())

# Display the results
print('Number of strategies:', len(strategy_counts))
print('Strategy counts:', strategy_counts)
print('Player strategies:', player_strategies)

Number of strategies: 9
Strategy counts: Counter({'Other': 118, 'Simpleton': 17, 'Always TRUST': 14, 'Always CHEAT': 13, 'Trust3 Cheat1': 11, 'COPYkitten': 10, 'CopyCat': 9, 'Grudger': 2, 'Reverse Copycat ': 2})
Player strategies: {174: 'CopyCat', 121: 'Always TRUST', 62: 'Other', 102: 'Other', 66: 'Other', 157: 'Other', 16: 'Other', 148: 'Other', 8: 'Trust3 Cheat1', 144: 'Other', 73: 'Other', 6: 'Trust3 Cheat1', 153: 'Other', 43: 'Trust3 Cheat1', 115: 'Always CHEAT', 70: 'Simpleton', 28: 'Other', 154: 'Simpleton', 31: 'Trust3 Cheat1', 83: 'Other', 71: 'Other', 32: 'Other', 69: 'Other', 98: 'Other', 54: 'Other', 182: 'Other', 50: 'Other', 7: 'Other', 100: 'COPYkitten', 49: 'Other', 137: 'Always TRUST', 96: 'Other', 89: 'Other', 113: 'CopyCat', 26: 'Other', 108: 'Always CHEAT', 22: 'Other', 47: 'Simpleton', 82: 'Simpleton', 149: 'COPYkitten', 35: 'Simpleton', 56: 'Other', 118: 'COPYkitten', 162: 'Other', 143: 'Other', 94: 'Other', 88: 'Other', 27: 'Other', 78: 'Other', 109: 'Other', 139

The above code can be used to identify many strategies like - 

1) Always Trust - In this strategy the player always plays 'TRUST' irrespective of the opponent's move.

2) Always Cheat - This strategy is the converse of the above strategy where the player always plays 'TRUST'. Here the player plays 'CHEAT' move only.

3) CopyCat - In this strategy the player imitates the opponent's last move.

4) Reverse Copycat - In this strategy the player plays the opposite move of the opponent's last move.

5) Simpleton - In this strategy the player imitates his own last move if the opponent's last move is 'TRUST' otherwise the player plays the opposite move of his own last move.

6) Grudger - In this strategy the player plays 'TRUST' until the opponent has played 'CHEAT' atleast once. After which only 'CHEAT' is played by player.

7) CopyKitten - Same as CopyCat just that the player cheats after the opponent has played 'CHEAT' twice consecutively.

8) Trust 3 Cheat 1 - In this the player plays 'TRUST' thrice and 'CHEAT' once in loop.

This is the data which we got according to the above used code:

Number of strategies: 9
Strategy counts: Counter({'Other': 118, 'Simpleton': 17, 'Always TRUST': 14, 'Always CHEAT': 13, 'Trust3 Cheat1': 11, 'COPYkitten': 10, 'CopyCat': 9, 'Grudger': 2, 'Reverse Copycat ': 2})
Player strategies: {1: 'Other', 2: 'COPYkitten', 3: 'COPYkitten', 4: 'Always CHEAT', 5: 'Other', 6: 'Trust3 Cheat1', 7: 'Other', 8: 'Trust3 Cheat1', 9: 'Always TRUST', 10: 'Other', 11: 'Other', 12: 'Other', 13: 'Always TRUST', 14: 'Grudger', 15: 'Always CHEAT', 16: 'Other', 17: 'Simpleton', 18: 'Other', 19: 'Other', 20: 'COPYkitten', 21: 'Other', 22: 'Other', 23: 'Other', 24: 'Other', 25: 'Always CHEAT', 26: 'Other', 27: 'Other', 28: 'Other', 29: 'Other', 30: 'CopyCat', 31: 'Trust3 Cheat1', 32: 'Other', 33: 'Always CHEAT', 34: 'Other', 35: 'Simpleton', 36: 'Other', 37: 'Other', 38: 'Other', 39: 'Other', 40: 'Other', 41: 'Other', 43: 'Trust3 Cheat1', 45: 'Simpleton', 46: 'Simpleton', 47: 'Simpleton', 48: 'Other', 49: 'Other', 50: 'Other', 51: 'Simpleton', 52: 'Simpleton', 53: 'Other', 54: 'Other', 55: 'Other', 56: 'Other', 57: 'Other', 58: 'Other', 59: 'Other', 60: 'Always TRUST', 61: 'Trust3 Cheat1', 62: 'Other', 63: 'Simpleton', 64: 'Other', 65: 'Other', 66: 'Other', 67: 'CopyCat', 68: 'Other', 69: 'Other', 70: 'Simpleton', 71: 'Other', 72: 'Other', 73: 'Other', 74: 'Other', 75: 'Other', 76: 'Other', 77: 'CopyCat', 78: 'Other', 80: 'Other', 81: 'Other', 82: 'Simpleton', 83: 'Other', 84: 'Other', 85: 'Other', 86: 'Other', 87: 'Other', 88: 'Other', 89: 'Other', 90: 'Other', 91: 'Other', 92: 'Other', 93: 'CopyCat', 94: 'Other', 95: 'Always TRUST', 96: 'Other', 97: 'Other', 98: 'Other', 99: 'Trust3 Cheat1', 100: 'COPYkitten', 101: 'Other', 102: 'Other', 103: 'Other', 104: 'Always TRUST', 105: 'Grudger', 106: 'Other', 107: 'Simpleton', 108: 'Always CHEAT', 109: 'Other', 110: 'Always CHEAT', 111: 'Always CHEAT', 112: 'Other', 113: 'CopyCat', 114: 'Other', 115: 'Always CHEAT', 116: 'Other', 117: 'Other', 118: 'COPYkitten', 119: 'Reverse Copycat ', 120: 'Other', 121: 'Always TRUST', 122: 'Other', 123: 'Other', 125: 'Other', 126: 'COPYkitten', 127: 'Trust3 Cheat1', 128: 'Always TRUST', 129: 'Always CHEAT', 130: 'Other', 131: 'Other', 132: 'Other', 133: 'Trust3 Cheat1', 134: 'Simpleton', 135: 'Other', 136: 'Always CHEAT', 137: 'Always TRUST', 138: 'COPYkitten', 139: 'Trust3 Cheat1', 140: 'Other', 141: 'Other', 142: 'Other', 143: 'Other', 144: 'Other', 145: 'Other', 146: 'Other', 147: 'Other', 148: 'Other', 149: 'COPYkitten', 150: 'COPYkitten', 151: 'Other', 152: 'Other', 153: 'Other', 154: 'Simpleton', 155: 'Always CHEAT', 156: 'Other', 157: 'Other', 158: 'Always CHEAT', 159: 'CopyCat', 160: 'Reverse Copycat ', 162: 'Other', 163: 'Other', 164: 'Other', 165: 'Simpleton', 166: 'Other', 167: 'Other', 168: 'Other', 169: 'Other', 170: 'Other', 171: 'Other', 172: 'Always CHEAT', 173: 'Other', 174: 'CopyCat', 175: 'Always TRUST', 176: 'Other', 177: 'Simpleton', 178: 'Always TRUST', 179: 'CopyCat', 180: 'CopyCat', 181: 'Always TRUST', 182: 'Other', 183: 'Other', 184: 'Other', 185: 'Other', 186: 'Trust3 Cheat1', 187: 'Other', 188: 'Other', 189: 'Other', 190: 'COPYkitten', 191: 'Trust3 Cheat1', 192: 'Always TRUST', 193: 'Always TRUST', 194: 'Always TRUST', 195: 'Simpleton', 196: 'Other', 197: 'Other', 198: 'Other', 199: 'Other', 200: 'Other', 201: 'Simpleton'}